src/hera/_cli/generate/hera.py

```diff
@@ -0,0 +1,434 @@
+    for field in hera_class.__fields__:
+        if field in template_keys and getattr(model_class_obj, field) is not None:
+            val, sub_hera_imports, sub_model_imports = python_obj_to_repr(getattr(model_class_obj, field))
+            hera_imports.extend(sub_hera_imports)
+            model_imports.extend(sub_model_imports)
+            hera_template_str.append(f"{field}={val},")
+
+    hera_template_str.append(")")
+    return (
+        "\n".join(hera_template_str),
+        hera_imports,
+        model_imports,
+    )
+
+
+def convert_to_hera_invocator_template(
+    template: Template,
+    hera_template_class: Type[BaseModel],
+    template_type: Optional[BaseModel] = None,
+) -> Tuple[str, List[str], List[str]]:
+    hera_imports = [hera_template_class.__name__]
+    model_imports = []
+
+    template_keys = set(hera_template_class.__fields__.keys()).intersection(template.__fields__.keys())
+
+    hera_template_str = [f"with {hera_template_class.__name__}("]
+    for field in template.__fields__:
+        if field in template_keys and getattr(template, field) is not None:
+            val, sub_hera_imports, sub_model_imports = python_obj_to_repr(getattr(template, field))
+            hera_imports.extend(sub_hera_imports)
+            model_imports.extend(sub_model_imports)
+            hera_template_str.append(f"{field}={val},")
+
+    if template_type:
+        template_field_keys = set(hera_template_class.__fields__.keys()).intersection(template_type.__fields__.keys())
+
+        for field in template_type.__fields__:
+            if hera_template_class == DAG and field == "tasks":
+                # We create Task objects within the context
+                continue
+            elif hera_template_class == ContainerSet and field == "containers":
+                continue
+
+            if field in template_field_keys and getattr(template_type, field) is not None:
+                val, sub_hera_imports, sub_model_imports = python_obj_to_repr(getattr(template_type, field))
+                hera_imports.extend(sub_hera_imports)
+                model_imports.extend(sub_model_imports)
+                hera_template_str.append(f"{field}={val},")
+
+    hera_template_str.append(") as invocator:")
+
+    indent = " " * 8
+
+    if hera_template_class == Steps and template.steps:
+        for parallel_steps in template.steps:
+            parallel_steps_list = parallel_steps.__root__
+            if len(parallel_steps_list) > 1:
+                hera_template_str.append(f"{indent}with invocator.parallel():")
+                indent = " " * 12
+            for step in parallel_steps.__root__:
+                val, sub_hera_imports, sub_model_imports = convert_to_hera_equivalent(step, Step)
+                hera_imports.extend(sub_hera_imports)
+                model_imports.extend(sub_model_imports)
+                hera_template_str.append(f"{indent}{val}")
+            indent = " " * 8
```

Comment:
I'm worried this pattern is fragile. We're relying on `convert_to_hera_equivalent` only emitting a single function call, for instance, since we're only indenting the start of the block. I'd prefer to create a class encapsulating "adding code + imports" and "do an indented block", and pass a mutable instance into the various functions. Something more like:

```python
code_writer.add_import(hera_template_class)
code_writer.add_line("with invocator.parallel():")
with code_writer.indent(4) as with_block:
    write_hera_equivalent(block, task, Task)
```

I'm aware this is a significant rewrite, though, so up to you if/when you want to do it.

Reply:
Yes I agree it's not the best. I started adding the `FileBuilder` to the code below, but didn't want to refactor this bit 😅 I like the `with_block` though, I intended to do something similar.